agent/csi/volumes.go

```diff
@@ -0,0 +1,248 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu        sync.RWMutex                     // To sync map "m" and "pluginMap"
+	tryMu     sync.RWMutex                     // To sync between tryAddVolume() and tryRemoveVolume()
+	m         map[string]*api.VolumeAssignment // Map between VolumeID and VolumeAssignment
+	pluginMap map[string]*NodePlugin           // Map between Driver Name and NodePlugin
+}
+
+const maxRetries int = 20
```

Comment:
Instead of just reducing `maxRetries`, can you clamp the maximum retry time at something 5 minutes? That seems short, but any longer than that and swarm ought to just reschedule the task.